// dbal/build/sql.js
"use strict";
class Sql {
    constructor(connection) {
        this.connection = connection;
    }
    static format(strings, ...params) {
        let query = strings.slice(0, 1);
        let values = [];
        params.forEach(function (value, i) {
            let append = strings[i + 1] || '';
            let left = strings[i].slice(-1);
            let right = append.slice(0, 1);
            if ((left === '"' && right === '"') || (left === "'" && right === "'")) {
                value = left === '"' ? value.replace(/"/g, '\\"') : value.replace(/'/g, "\\'");
            }
            else {
                value = '$' + values.push(value);
            }
            query.push(value, append);
        });
        let text = query.join('').trim();
        return { text, values };
    }
    format(strings, ...params) {
        return Sql.format(strings, ...params);
    }
    run(strings, ...params) {
        return this.connection.run(Sql.format(strings, ...params));
    }
    all(strings, ...params) {
        return this.connection.all(Sql.format(strings, ...params));
    }
    one(strings, ...params) {
        return this.connection.one(Sql.format(strings, ...params));
    }
}
exports.Sql = Sql;
